src/Architecture.cpp

```diff
@@ -495,69 +517,56 @@ void Architecture::findCouplingLimit(unsigned short node, int curSum, const std:
     visited[node] = false;
 }
 
-void Architecture::getHighestFidelityCouplingMap(unsigned short nQubits, CouplingMap& reducedMap) {
-    if (architectureName.empty() || nqubits == nQubits ||
+void Architecture::getHighestFidelityCouplingMap(unsigned short subsetSize, CouplingMap& reducedMap) {
+    if (!isArchitectureAvailable() || nqubits == subsetSize ||
         calibrationName.empty()) {
         reducedMap = couplingMap;
     } else {
-        double              bestFidelity{};
-        std::vector<double> allFidelities{};
-        auto                allConnectedSubsets = getAllConnectedSubsets(nQubits);
+        double bestFidelity{};
+        double currentFidelity     = 0.0;
+        auto   allConnectedSubsets = getAllConnectedSubsets(subsetSize);
 
-        allFidelities.reserve(allConnectedSubsets.size());
         for (const auto& qubitChoice: allConnectedSubsets) {
             CouplingMap map{};
             getReducedCouplingMap(qubitChoice, map);
-            bestFidelity = getFidelity(map, qubitChoice, calibrationData);
-            if (allFidelities.empty()) {
-                allFidelities.emplace_back(bestFidelity);
-                reducedMap = map;
-            } else {
-                auto it = allFidelities.begin();
-                while (it != allFidelities.end() && *it < bestFidelity) {
-                    std::advance(it, 1);
-                }
-                const auto distance = std::abs(std::distance(allFidelities.begin(), it));
-                allFidelities.emplace(allFidelities.begin() + distance, bestFidelity);
-                if (distance == 0) {
-                    reducedMap = map;
-                }
+            bestFidelity = getAverageArchitectureFidelity(map, qubitChoice, calibrationData);
+            if (currentFidelity < bestFidelity) {
+                reducedMap      = map;
+                currentFidelity = bestFidelity;
             }
         }
     }
 }
-std::vector<std::set<unsigned short>> Architecture::getAllConnectedSubsets(unsigned short nQubits) {
+std::vector<std::set<unsigned short>> Architecture::getAllConnectedSubsets(unsigned short subsetSize) {
     std::vector<std::set<unsigned short>> result{};
-    if (architectureName.empty() || nqubits == nQubits) {
-        result.emplace_back();
-        for (unsigned short i = 0U; i < nQubits; ++i) {
-            result.back().emplace(i);
-        }
-    } else if (nqubits < nQubits) {
+    if (architectureName.empty() || nqubits == subsetSize) {
+        result.emplace_back(getQubitSet());
+    } else if (nqubits < subsetSize) {
         throw QMAPException("Architecture too small!");
     } else {
-        for (const auto& subset: subsets(getQubitSet(), nQubits)) {
-            if (isFullyConnected(couplingMap, static_cast<int>(nqubits), subset)) {
+        for (const auto& subset: subsets(getQubitSet(), subsetSize)) {
```

Comment:
The subsets routine is used in the CliffordOptimizer, with and without checking connectivity, at least once. Critique is noted and if in refactor of CliffordSynthesizer this is no longer necessary it will be removed

Reply:
wouldn't it still be worth to rewrite this loop at the cost of some code duplication?